{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5e58084f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Sequence length: 8\n",
      "Matching count: 6\n",
      "Percentage: 75.0%\n"
     ]
    }
   ],
   "source": [
    "# qsn 1\n",
    "seq1 = 'ATGCATAG'\n",
    "seq2 = 'ATGAATGG'\n",
    "\n",
    "def matching(seq1, seq2):\n",
    "    match = 0\n",
    "    # Check identity (identity = No of exacat matches/Total positions)*100 \n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] == seq2[i]:\n",
    "            match+=1\n",
    "\n",
    "    total = len(seq1)\n",
    "    percent_identity = (match/total)*100\n",
    "    print(f\"Total Sequence length: {total}\")\n",
    "    print(f\"Matching count: {match}\")\n",
    "    print(f\"Percentage: {percent_identity}%\")\n",
    "\n",
    "matching(seq1, seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "22e3b0df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcription of DNA Sequence1: UACCGACUUGAC\n",
      "Transcription of DNA Sequence1: UACCGGCUUAAC\n",
      "\n",
      "Codons of DNA Sequence1: ['UAC', 'CGA', 'CUU', 'GAC']\n",
      "Codons of DNA Sequence2: ['UAC', 'CGG', 'CUU', 'AAC']\n",
      "\n",
      "Amino sequence of DNA Sequence1: ['Tyr', 'Arg', 'Leu', 'Asp']\n",
      "Amino sequence of DNA Sequence2: ['Tyr', 'Arg', 'Leu', 'Asn']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# qsn 2\n",
    "DNA_seq1 = 'ATGGCTGAACTG'\n",
    "DNA_seq2 = 'ATGGCCGAATTG'\n",
    "\n",
    "# Transcription\n",
    "def transcript(sequence):\n",
    "    mrna = ''\n",
    "    for i in sequence:\n",
    "        if i == 'A':\n",
    "            mrna += 'U'\n",
    "        elif i == 'T':\n",
    "            mrna += 'A'   \n",
    "        elif i == 'C':\n",
    "            mrna += 'G'  \n",
    "        else:\n",
    "            mrna += 'C'\n",
    "    return mrna\n",
    "    \n",
    "# Codons\n",
    "def group_letters(mrna):\n",
    "    groups = []\n",
    "    for i in range(0, len(mrna), 3):\n",
    "        groups.append(mrna[i:i+3])\n",
    "    return groups\n",
    "\n",
    "# Translation (Protein)\n",
    "def protein(groups):\n",
    "    amino = []\n",
    "    for i in groups:\n",
    "        if i in ['UUU', 'UUC']:\n",
    "            amino.append('Phe')\n",
    "        elif i in ['UUA', 'UUG', 'CUU', 'CUC', 'CUA', 'CUG']:\n",
    "            amino.append('Leu')\n",
    "        elif i in ['AUU', 'AUC', 'AUA']:\n",
    "            amino.append('Ile')\n",
    "        elif i == 'AUG':\n",
    "            amino.append('Met')\n",
    "        elif i in ['GUU', 'GUC', 'GUA', 'GUG']:\n",
    "            amino.append('Val')\n",
    "        elif i in ['UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC']:\n",
    "            amino.append('Ser')\n",
    "        elif i in ['CCU', 'CCC', 'CCA', 'CCG']:\n",
    "            amino.append('Pro')\n",
    "        elif i in ['ACU', 'ACC', 'ACA', 'ACG']:\n",
    "            amino.append('Thr')\n",
    "        elif i in ['GCU', 'GCC', 'GCA', 'GCG']:\n",
    "            amino.append('Ala')\n",
    "        elif i in ['UAU', 'UAC']:\n",
    "            amino.append('Tyr')\n",
    "        elif i in ['UAA', 'UAG', 'UGA']:\n",
    "            amino.append('Stop')\n",
    "        elif i in ['CAU', 'CAC']:\n",
    "            amino.append('His')\n",
    "        elif i in ['CAA', 'CAG']:\n",
    "            amino.append('Gln')\n",
    "        elif i in ['AAU', 'AAC']:\n",
    "            amino.append('Asn')\n",
    "        elif i in ['AAA', 'AAG']:\n",
    "            amino.append('Lys')\n",
    "        elif i in ['GAU', 'GAC']:\n",
    "            amino.append('Asp')\n",
    "        elif i in ['GAA', 'GAG']:\n",
    "            amino.append('Glu')\n",
    "        elif i in ['UGU', 'UGC']:\n",
    "            amino.append('Cys')\n",
    "        elif i == 'UGG':\n",
    "            amino.append('Trp')\n",
    "        elif i in ['CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG']:\n",
    "            amino.append('Arg')\n",
    "        elif i in ['GGU', 'GGC', 'GGA', 'GGG']:\n",
    "            amino.append('Gly')\n",
    "        else:\n",
    "            amino.append('??? - ')\n",
    "    return amino\n",
    "    \n",
    "    \n",
    "transcript_1 = transcript(DNA_seq1)\n",
    "print(f\"Transcription of DNA Sequence1: {transcript_1}\")\n",
    "\n",
    "transcript_2 = transcript(DNA_seq2)\n",
    "print(f\"Transcription of DNA Sequence1: {transcript_2}\\n\")\n",
    "\n",
    "codon_1 = group_letters(transcript_1)\n",
    "print(f\"Codons of DNA Sequence1: {codon_1}\")\n",
    "codon_2 = group_letters(transcript_2)\n",
    "print(f\"Codons of DNA Sequence2: {codon_2}\\n\")\n",
    "\n",
    "translation_1 = protein(codon_1)\n",
    "print(f\"Amino sequence of DNA Sequence1: {translation_1}\")\n",
    "translation_2 = protein(codon_2)\n",
    "print(f\"Amino sequence of DNA Sequence2: {translation_2}\\n\")\n",
    "\n",
    "# Amino acid similarity\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "fc0a6b3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Substitution:  ATGCCATAGCTAAGGCTT\n",
      "Insertion:  ATGCCAGTAGCTAAGGCTT\n",
      "Deletion:  ATGCCTAGCTAAGGCTT\n",
      "\n",
      "\n",
      "Amino sequence of Original DNA Sequence: ['Tyr', 'Gly', 'Ile', 'Asp', 'Ser', 'Glu']\n",
      "Amino sequence of Substitution Sequence: ['Tyr', 'Gly', 'Ile', 'Asp', 'Ser', 'Glu']\n",
      "Amino sequence of Insertion Sequence: ['Tyr', 'Gly', 'His', 'Arg', 'Phe', 'Arg', '??? - ']\n",
      "Amino sequence of Deletion Sequence: ['Tyr', 'Gly', 'Ser', 'Ile', 'Pro', '??? - ']\n",
      "\n",
      "Original vs Substituted: \n",
      "Total Sequence length: 6\n",
      "Matching count: 6\n",
      "Percentage: 100.0%\n",
      "\n",
      "Original vs Insertion: \n",
      "Total Sequence length: 6\n",
      "Matching count: 2\n",
      "Percentage: 33.33333333333333%\n",
      "\n",
      "Original vs Deletion: \n",
      "Total Sequence length: 6\n",
      "Matching count: 2\n",
      "Percentage: 33.33333333333333%\n"
     ]
    }
   ],
   "source": [
    "# DNA Seq = \"ATGCCGTAGCTAAGGCTT\"\n",
    "    # Create a substitution mutation\n",
    "    # An Insertion mutation\n",
    "    # A deletion mutation\n",
    "# Translate all 4 seq into protein\n",
    "# Compare all protein lengths\n",
    "# Compare original dna identity pctg with mutated seq\n",
    "\n",
    "DNA_Seq = \"ATGCCGTAGCTAAGGCTT\"\n",
    "\n",
    "# Substitution\n",
    "sub_seq = DNA_Seq[:5] + \"A\" + DNA_Seq[6:]\n",
    "print(\"Substitution: \", sub_seq)\n",
    "\n",
    "# Insertion\n",
    "in_seq = DNA_Seq[:5] + \"A\" + DNA_Seq[5:]\n",
    "print(\"Insertion: \", in_seq)\n",
    "\n",
    "# Deletion\n",
    "del_seq = DNA_Seq[:5] + DNA_Seq[6:]\n",
    "print(\"Deletion: \", del_seq)\n",
    "\n",
    "# Translation to Protein\n",
    "print('\\n')\n",
    "transcript_1 = transcript(DNA_Seq)\n",
    "transcript_2 = transcript(sub_seq)\n",
    "transcript_3 = transcript(in_seq)\n",
    "transcript_4 = transcript(del_seq)\n",
    "\n",
    "codon_1 = group_letters(transcript_1)\n",
    "codon_2 = group_letters(transcript_2)\n",
    "codon_3 = group_letters(transcript_3)\n",
    "codon_4 = group_letters(transcript_4)\n",
    "\n",
    "translation_1 = protein(codon_1)\n",
    "print(f\"Amino sequence of Original DNA Sequence: {translation_1}\")\n",
    "translation_2 = protein(codon_2)\n",
    "print(f\"Amino sequence of Substitution Sequence: {translation_2}\")\n",
    "translation_3 = protein(codon_3)\n",
    "print(f\"Amino sequence of Insertion Sequence: {translation_3}\")\n",
    "translation_4 = protein(codon_4)\n",
    "print(f\"Amino sequence of Deletion Sequence: {translation_4}\\n\")\n",
    "\n",
    "print(\"Original vs Substituted: \")\n",
    "matching(translation_1, translation_2)\n",
    "\n",
    "print(\"\\nOriginal vs Insertion: \")\n",
    "matching(translation_1, translation_3)\n",
    "\n",
    "print(\"\\nOriginal vs Deletion: \")\n",
    "matching(translation_1, translation_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1b3edd35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteins:\n",
      "Original: MP Length: 2\n",
      "Substitution: MP Length: 2\n",
      "Insertion: MPVAKA Length: 6\n",
      "Deletion: MPS Length: 3\n",
      "\n",
      "DNA Identity Percentages:\n",
      "Original vs Substitution: 94.44444444444444\n",
      "Original vs Insertion: 42.10526315789473\n",
      "Original vs Deletion: 44.44444444444444\n"
     ]
    }
   ],
   "source": [
    "# Manual DNA mutation and translation\n",
    "\n",
    "DNA_Seq = \"ATGCCGTAGCTAAGGCTT\"\n",
    "\n",
    "# --- Mutations ---\n",
    "# Substitution: replace base at position 5 with 'A'\n",
    "sub_seq = DNA_Seq[:5] + \"A\" + DNA_Seq[6:]\n",
    "\n",
    "# Insertion: insert 'T' after position 5\n",
    "ins_seq = DNA_Seq[:5] + \"A\" + DNA_Seq[5:]\n",
    "\n",
    "# Deletion: remove base at position 5\n",
    "del_seq = DNA_Seq[:5] + DNA_Seq[6:]\n",
    "\n",
    "# --- Codon to Amino Acid Table (simplified) ---\n",
    "codon_table = {\n",
    "    \"ATA\":\"I\",\"ATC\":\"I\",\"ATT\":\"I\",\"ATG\":\"M\",\n",
    "    \"ACA\":\"T\",\"ACC\":\"T\",\"ACG\":\"T\",\"ACT\":\"T\",\n",
    "    \"AAC\":\"N\",\"AAT\":\"N\",\"AAA\":\"K\",\"AAG\":\"K\",\n",
    "    \"AGC\":\"S\",\"AGT\":\"S\",\"AGA\":\"R\",\"AGG\":\"R\",\n",
    "    \"CTA\":\"L\",\"CTC\":\"L\",\"CTG\":\"L\",\"CTT\":\"L\",\n",
    "    \"CCA\":\"P\",\"CCC\":\"P\",\"CCG\":\"P\",\"CCT\":\"P\",\n",
    "    \"CAC\":\"H\",\"CAT\":\"H\",\"CAA\":\"Q\",\"CAG\":\"Q\",\n",
    "    \"CGA\":\"R\",\"CGC\":\"R\",\"CGG\":\"R\",\"CGT\":\"R\",\n",
    "    \"GTA\":\"V\",\"GTC\":\"V\",\"GTG\":\"V\",\"GTT\":\"V\",\n",
    "    \"GCA\":\"A\",\"GCC\":\"A\",\"GCG\":\"A\",\"GCT\":\"A\",\n",
    "    \"GAC\":\"D\",\"GAT\":\"D\",\"GAA\":\"E\",\"GAG\":\"E\",\n",
    "    \"GGA\":\"G\",\"GGC\":\"G\",\"GGG\":\"G\",\"GGT\":\"G\",\n",
    "    \"TCA\":\"S\",\"TCC\":\"S\",\"TCG\":\"S\",\"TCT\":\"S\",\n",
    "    \"TTC\":\"F\",\"TTT\":\"F\",\"TTA\":\"L\",\"TTG\":\"L\",\n",
    "    \"TAC\":\"Y\",\"TAT\":\"Y\",\"TAA\":\"STOP\",\"TAG\":\"STOP\",\n",
    "    \"TGC\":\"C\",\"TGT\":\"C\",\"TGA\":\"STOP\",\"TGG\":\"W\",\n",
    "}\n",
    "\n",
    "# --- Translation Function ---\n",
    "def translate_dna(seq):\n",
    "    protein = \"\"\n",
    "    for i in range(0, len(seq), 3):\n",
    "        codon = seq[i:i+3]\n",
    "        if len(codon) < 3:\n",
    "            break\n",
    "        aa = codon_table.get(codon, \"\")\n",
    "        if aa == \"STOP\":\n",
    "            break\n",
    "        protein += aa\n",
    "    return protein\n",
    "\n",
    "# --- Translate all sequences ---\n",
    "original_protein = translate_dna(DNA_Seq)\n",
    "sub_protein = translate_dna(sub_seq)\n",
    "ins_protein = translate_dna(ins_seq)\n",
    "del_protein = translate_dna(del_seq)\n",
    "\n",
    "# --- Identity Percentage ---\n",
    "def identity_percentage(seq1, seq2):\n",
    "    matches = sum(1 for a, b in zip(seq1, seq2) if a == b)\n",
    "    return matches / max(len(seq1), len(seq2)) * 100\n",
    "\n",
    "# --- Results ---\n",
    "print(\"Proteins:\")\n",
    "print(\"Original:\", original_protein, \"Length:\", len(original_protein))\n",
    "print(\"Substitution:\", sub_protein, \"Length:\", len(sub_protein))\n",
    "print(\"Insertion:\", ins_protein, \"Length:\", len(ins_protein))\n",
    "print(\"Deletion:\", del_protein, \"Length:\", len(del_protein))\n",
    "\n",
    "print(\"\\nDNA Identity Percentages:\")\n",
    "print(\"Original vs Substitution:\", identity_percentage(DNA_Seq, sub_seq))\n",
    "print(\"Original vs Insertion:\", identity_percentage(DNA_Seq, ins_seq))\n",
    "print(\"Original vs Deletion:\", identity_percentage(DNA_Seq, del_seq))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
